### PR TITLE
JDK-8289524: Add JFR JIT restart event

### PR DESCRIPTION
**Pull request by MBaesken**
The JIT compiler restarts (see restart_compiler in NMethodSweeper::sweep_code_cache) would be a helpful addition to the JFR events. Currently we log the JIT stop operations in JFR (EventCodeCacheFull) but no restart.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289524](https://bugs.openjdk.org/browse/JDK-8289524): Add JFR JIT restart event

### Reviewers
 * @parttimenerd (no known github.com user name / role) ⚠️ Review applies to [29665ecb](https://git.openjdk.org/jdk/pull/9334/files/29665ecb9f3a86077fcc564daacc8a5cd205568e)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [29665ecb](https://git.openjdk.org/jdk/pull/9334/files/29665ecb9f3a86077fcc564daacc8a5cd205568e)
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9334/head:pull/9334` \
`$ git checkout pull/9334`

Update a local copy of the PR: \
`$ git checkout pull/9334` \
`$ git pull https://git.openjdk.org/jdk pull/9334/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9334`

View PR using the GUI difftool: \
`$ git pr show -t 9334`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9334.diff">https://git.openjdk.org/jdk/pull/9334.diff</a>

</details>

### CODE REVIEW

**Review by parttimenerd — APPROVED** [review]
Looks good and simple

**Comment by MBaesken** [comment]
> Looks good and simple

Hi, thanks for the review ! May I have a second review ?

**Comment by mgronlun** [comment]
Thanks for clarifying Erik, I forgot to mention the contentType.

**Comment by MBaesken** [comment]
Hi Markus,  I think it would be possible and probably make sense to instead enhance the existing EventSweepCodeCache.
Would it be okay to move down the EventSweepCodeCache to the end of NMethodSweeper::sweep_code_cache()  ?
I think post_sweep_event  would get  2 more parameters, a boolean for JIT-restart and an integer freed_memory.
Are you fine with this ?
Are there some compatibility concerns when enhancing an existing event EventSweepCodeCache ?

**Comment by mgronlun** [comment]
Thanks for considering. It is fine to move the EventSweepCodeCache event around, you can see that the constructor takes the UNTIMED value. This means that timestamping is handled explicitly, by setting the set_starttime() and set_endtime() fields For compatibility, we can think of it like we are creating a subclass with extended fields. Just append the additional fields to the event. As for the memory field declarations, you can take a peek at other events in metadata.xml, for example, this field declaration is from CodeCacheFull:

    <Field type="ulong" contentType="bytes" name="unallocatedCapacity" label="Unallocated" />

**Comment by MBaesken** [comment]
Hi Markus, another question that came up while looking into this - the current SweepCodeCache event has a threshold of 100 ms set in both default.jfc and profile.jfc .  This was probably fine for existing usages.  But would we loose the JIT restart events sometimes in case we incorporate the Jit restart and frred memory into the current SweepCodeCache   event ?

**Comment by mgronlun** [comment]
Can you explain a bit more about what "JIT restart" actually means?

**Comment by MBaesken** [comment]
> Can you explain a bit more about what "JIT restart" actually means?

The comment at line 433 of sweeper.cpp and following is explaining it.
The JIT compilation had been stopped before and now, after sweeping freed potentially some memory,
JIT compilation  is re-enabled.
That's why also the  log.debug("restart compiler");  in the code.

**Comment by mgronlun** [comment]
Ok, so there is a corresponding "compiler stopped", implicitly noted by firing EventCodeCacheFull?

**Comment by mgronlun** [comment]
Hi again Mathias, based on your observation that this does not really align well with the threshold parameter for SweepCodeCache and is a complement event to EventCodeCacheFull, I believe your original suggestion is better (an instant event, only for "jit restart"). I was slightly tripped up with the concept of a  "JIT restart", because at this point the JIT is already stopped, as implied by EventCodeCacheFull. So "JIT restart" here contextually means, "JIT start", as in "start JITting code again". Do you know if we have an event that describes the CodeCache settings in terms of memory sizes so that the "freed memory" can be interpreted relative to it?

**Review by parttimenerd — APPROVED** [review]

**Comment by egahlin** [comment]
It would be good to add a sanity test.  See:
https://github.com/openjdk/jdk/blob/83a5d5996bca26b5f2e97b67f9bfd0a6ad110327/test/jdk/jdk/jfr/event/compiler/TestCodeSweeper.java#L181

The test in on the ProblemList.txt due to timeouts, but probably works most of the time. 

Also remove ""JitRestart" from TestLookForUntestedEvents.java

**Comment by MBaesken** [comment]
> Yes, thank you, I took a look at those fields too. Unfortunately, there is no value that bears directly on "current in-use". The reserved and committed will not be updated. If there was a value that could reflect the current usage, even if it means adding it to both EventCodeCacheFull and the JIT restart, that would be perfect. EventCodeCacheFull because in-use is x, JIT restart because in-use is now y. There are some statistics-related properties in the CodeHeap, like for example heap->unallocated_capacity(); et al. Could some of those expose this running value perhaps?

Probably we would need  CodeCache::unallocated_capacity()  (contains a number of code heaps as far as I know)  to compare freed_memory to, because from what I see we iterate the whole CodeCache in NMethodSweeper::sweep_code_cache() .
But maybe someone with better knowledge about the CodeCache should comment here.

**Comment by RealLucy** [comment]
To get an idea of how much memory was freed in relation to what's actually available, you could evaluate CodeCache:max_capacity(). This function returns the overall size of the code heap (across all segments). The performance impact is minimal and no locks are acquired.

**Comment by MBaesken** [comment]
So is the common opinion to get back to a separate JIT start  event  (I think that naming is prefered over JIT restart, am I correct) ?
Additionally we add CodeCache:max_capacity() at both JIT start  and  EventCodeCacheFull ?

**Comment by RealLucy** [comment]
> So is the common opinion to get back to a separate JIT start event (I think that naming is prefered over JIT restart, am I correct) ?
> Additionally we add CodeCache:max_capacity() at both JIT start and EventCodeCacheFull ?

I would be in favour of that.

**Comment by mgronlun** [comment]
> So is the common opinion to get back to a separate JIT start event (I think that naming is prefered over JIT restart, am I correct) ? Additionally we add CodeCache:max_capacity() at both JIT start and EventCodeCacheFull ?

I think it is fine to use the term "JIT restart" because it is in use both in the code as well as in the output of log statements. Another reason would be that the first "JIT start" event would always be missing. It was only the meaning that got me a bit confused. Yes, a separate event "JIT is preferrable having no duration (startTime=false), no stack trace(stackTrace=false). The .jfc configs only need one element:

<setting name="enabled" control="compiler-enabled">true</setting>

**Review by RealLucy — APPROVED** [review]
Changes look good. 
Now you can relate the freed memory to what's available in total.

File: src/hotspot/share/jfr/metadata/metadata.xml
Comment (mgronlun) [inline src/hotspot/share/jfr/metadata/metadata.xml:563]: Is codeCacheMaxCapacity the current in-use size of the CodeCache, in bytes? It should have the contentType="bytes" in that case. Also "freedMemory" should have the same contentType.

File: src/hotspot/share/jfr/metadata/metadata.xml
Comment (mgronlun) [inline src/hotspot/share/jfr/metadata/metadata.xml:620]: contentType="bytes"

File: test/jdk/jdk/jfr/event/metadata/TestLookForUntestedEvents.java
Comment (mgronlun) [inline test/jdk/jdk/jfr/event/metadata/TestLookForUntestedEvents.java:57]: Can we create a test? There is an existing test for CodeCacheFull, maybe derive from it to create a test also for the JitRestart event?

**Review by mgronlun — CHANGES_REQUESTED** [review]

File: src/hotspot/share/jfr/metadata/metadata.xml
Comment (egahlin) [inline src/hotspot/share/jfr/metadata/metadata.xml:561]: "JITRestart" . The convention for other events have been to use capital laters for acronyms. 

"JIT restart" -> "JIT Restart"

File: src/hotspot/share/code/codeCache.cpp
Comment (egahlin) [inline src/hotspot/share/code/codeCache.cpp:1369]: Add test of field in TestCodeCacheFull

File: src/hotspot/share/runtime/sweeper.cpp
Comment (egahlin) [inline src/hotspot/share/runtime/sweeper.cpp:440]: It would be good to have a sanity test of the event. 

If the event can't be provoked reliably, retry or accept as OK.

**Comment by MBaesken** [comment]
Hi , I adjusted the contentType as suggested, adjusted the label  and TestCodeCacheFull.
Regarding triggering the JitRestart event , the  suggested TestCodeCacheFull looks like a good start . So I  added to allocateCodeBlob (this triggers the CodeCacheFull event)  a  corresponding freeCodeBlob call .
Unfortunately this did not work in practise and doesn't show the JitRestart event.  There could be some delay to the restarting unfortunately.

**Comment by MBaesken** [comment]
I added a test after some playing around with the WhiteBox functionality.

File: test/jdk/jdk/jfr/event/compiler/TestJitRestart.java
Comment (coleenp) [inline test/jdk/jdk/jfr/event/compiler/TestJitRestart.java:33]: This package has been removed. Please use jdk.test.whitebox.WhiteBox.

File: src/hotspot/share/jfr/metadata/metadata.xml
Comment (RealLucy) [reply src/hotspot/share/jfr/metadata/metadata.xml]: Yes, codeCacheMaxCapacity is given in bytes. It specifies the maximum size, not the current in-use size, of the CodeCache. It is the sum over all CodeHeap segments.

File: test/jdk/jdk/jfr/event/compiler/TestJitRestart.java
Comment (MBaesken) [reply test/jdk/jdk/jfr/event/compiler/TestJitRestart.java]: Hi Coleen thanks for the advice.  After switching to the new package  jdk.test.whitebox.WhiteBox  I get 
java.lang.UnsatisfiedLinkError: 'void sun.hotspot.WhiteBox.registerNatives()'
Do I need to do more than just renaming the package ?

Comment (MBaesken) [reply test/jdk/jdk/jfr/event/compiler/TestJitRestart.java]: Had to change BlobType import  too, this one is also in a new package.

**Review by RealLucy — APPROVED** [review]
With all the fine-tuning, changes look even better now. And there is a test now so we get alerted if anything breaks. Thanks!

**Pull request closed without merging**
